qiskit/providers/aer/noise/passes.py

```diff
@@ -0,0 +1,209 @@
+from .errors import QuantumError
+from .errors.standard_errors import thermal_relaxation_error
+
+InstructionLike = Union[Instruction, QuantumError]
+
+
+class LocalNoisePass(TransformationPass):
+    """Transpiler pass to insert noise into a circuit.
+
+    The noise in this pass is defined by a noise function or callable with signature
+
+    .. code:: python
+
+            def fn(
+                inst: Instruction,
+                qubits: Optional[List[int]] = None
+            ) -> InstructionLike:
+
+    For every instance of one of the reference instructions in a circuit the
+    supplied function is called on that instruction and the returned noise
+    is added to the circuit. This noise can depend on properties of the
+    instruction it is called on (for example parameters or duration) to
+    allow inserting parameterized noise models.
+
+    Several methods for adding the constructed errors to circuits are supported
+    and can be set by using the ``method`` kwarg. The supported methods are
+
+    * ``"append"``: add the return of the callable after the instruction.
+    * ``"prepend"``: add the return of the callable before the instruction.
+    * ``"replace"``: replace the instruction with the return of the callable.
+
+    """
+
+    def __init__(
+            self,
+            fn: Callable[[Instruction, Sequence[int]], InstructionLike],
+            ops: Optional[Union[Instruction, Iterable[Instruction]]] = None,
+            method: str = 'append'
+    ):
+        """Initialize noise pass.
+
+        Args:
+            fn: noise function `fn(inst, qubits) -> InstructionLike`.
+            ops: Optional, single or list of instructions to apply the
+                                  noise function to. If None the noise function will be
+                                  applied to all instructions in the circuit.
+            method: method for inserting noise. Allow methods are
+                           'append', 'prepend', 'replace'.
+        Raises:
+            TranspilerError: if an invalid option is specified.
+        """
+        if method not in {"append", "prepend", "replace"}:
+            raise TranspilerError(
+                f'Invalid method: {method}, it must be "append", "prepend" or "replace"'
+            )
+        if isinstance(ops, str):
+            ops = [ops]
+        super().__init__()
+        self._fn = fn
+        self._ops = set(ops) if ops else {}
+        self._method = method
+
+    def run(self, dag: DAGCircuit) -> DAGCircuit:
+        """Run the LocalNoisePass pass on `dag`.
+        Args:
+            dag: DAG to be changed.
+        Returns:
+            A changed DAG.
+        Raises:
+            TranspilerError: if generated operation is not valid.
+        """
+        qubit_indices = {qubit: idx for idx, qubit in enumerate(dag.qubits)}
+        for node in dag.topological_op_nodes():
+            if self._ops and node.op.name not in self._ops:
```

Comment:
I originally thought we should use instances of ops rather than names. Names strings for instructions can be a bit problematic because you can get situations where a circuit might end up with the same name as an instruction you added to the noise model.

The original motivation for using instances is so you could define diferent noise for cases of the same instruction by subclassing and doing something like

```
class CX1(CXGate):
    pass

class CX2(CXGate):
     pass
```
And add different errors to to the alternative versions of the same gate.

If you make this change then this code would look like something like
```
qubit_indices = {qubit: idx for idx, qubit in enumerate(dag.qubits)}
op_types = tuple(self._ops)
 for node in dag.topological_op_nodes():
    if self._ops and isinstance(node.op, op_types):
```

Reply:
Done at 5753075